core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -64,15 +65,8 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
       }
     }
 
-    Set<ManifestFile> candidateManifestFilesForDeletion = readManifests(expiredSnapshots);
     Set<ManifestFile> manifestFilesAfterExpiration = readManifests(snapshotsAfterExpiration);
-
-    Set<ManifestFile> manifestsToDelete = Sets.newHashSet();
-    for (ManifestFile candidateManifestFile : candidateManifestFilesForDeletion) {
-      if (!manifestFilesAfterExpiration.contains(candidateManifestFile)) {
-        manifestsToDelete.add(candidateManifestFile);
-      }
-    }
+    Set<ManifestFile> manifestsToDelete = manifestFilesToDelete(manifestFilesAfterExpiration, expiredSnapshots);
```

Comment:
With the task based approach we should guarantee if there's a failure that it gets thrown and doesn't proceed to the next step. I've made sure to throw in case determining manifests fails